Using a service as a router resource

weaverryan opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | almost |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | not yet... |

Hi guys!

This adds the ability to use a service as a routing resource. In other words, instead of loading `routing.yml`, you could load `my_route_loader`, and then a method would be called on your service to return a RouteCollection.

Specifically, I'm interested in this because it would allow a user to point their main router resource to the kernel itself, making it possible to load routes inside the kernel (making a single-file full-stack app more possible).

Thanks!

weaverryan committed:
Adding a new route loader that is able to load routes from a service

This makes it possible to list a service as your main routing "resource", making it
easily possible to load routes from within a Php class, such as the kernel itself.

weaverryan committed:
Registering the route loader as a service

weaverryan committed:
fabbot!

jakzal commented on src/Symfony/Bundle/FrameworkBundle/Routing/ServiceRouterLoader.php:
```diff
@@ -0,0 +1,71 @@
+     */
+    public function supports($resource, $type = null)
+    {
+        return $type == 'service';
```
Should be `return 'service' === $type`.

jakzal commented on src/Symfony/Bundle/FrameworkBundle/Tests/Routing/ServiceRouterLoaderTest.php:
```diff
@@ -0,0 +1,72 @@
+        $container = $this->getMock('Symfony\\Component\\DependencyInjection\\Container', array('get'));
+
+        $container
+            ->expects($this->once())
```
I wouldn't use `$this->once()` in case of stubs, `$this->any()` should be sufficient. The same applies to the other stub in this test.

We shouldn't care how many times this method is called. By relying on number of calls we make assumptions about implementation. All is important here that if a method is called, it will return a stubbed value. Assertions at the end perform verifications.

stof commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteLoaderInterface.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return RouteCollection
+     */
+    public function getRouteCollection(Loader $loader);
```
please don't typehint an implementation

weaverryan replied:
My problem is that the `Loader` has `import` on it, which is really what the user will want. The `LoaderInterface` doesn't have this. What do you think about creating a new interface for loaders with `import` to avoid type-hinting the implementation?

stof commented on src/Symfony/Bundle/FrameworkBundle/Routing/ServiceRouterLoader.php:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class ServiceRouterLoader extends Loader
```
should this really be in FrameworkBundle ? this makes the feature unusable for projects using the router and DI components without using the fullstack framework (Drupal is such a project for instance).

This could be an optional feature in the component

weaverryan replied:
I agree. I think this should be in the Routing component then - I'll make that change.

weaverryan committed:
Moving the ServiceRouteLoader stuff into the Routing component as an optional feature

weaverryan committed:
Some tweaks thanks to @jakzal

weaverryan committed:
Hinting against the interface

I kind of don't want this, because I'd rather promise the user that there is an "import"
function. But, I think it is feasible that someone wants to use this, but has simple
loaders that don't have the ability to import (i.e. don't extend Loader).

weaverryan commented:
@stof I decided to type-hint against the existing `LoaderInterface`, as it's the smallest change. I'd still love to have `import` guaranteed, but I believe you can "import" manually by calling `getResolver()->resolve(...)->load()` on the `LoaderInterface`. If there's support for adding some sort of an `ImportingLoaderInterface` with the `import()` method, I would be very happy to add that.

I've taken all the feedback into account. This is still ready for review :)

weaverryan committed:
Adding the entire class hierarchy as a resource

The addClassResource() was taken from ContainerBuilder

weaverryan commented on src/Symfony/Component/Routing/Loader/RouteLoaderInterface.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return RouteCollection
+     */
+    public function getRouteCollection(LoaderInterface $loader);
```
I've type-hinted the interface, but in reality, I'm passing in the concrete `ServiceRouteLoader` itself. The difference is that `ServiceRouteLoader` (like all loaders), has a really handy `import()` method, while `LoaderInterface` does not. I'd prefer to type-hint something with the `import()` method - either the concrete `Loader` or a new interface.

weaverryan commented:
Ping @symfony/deciders

weaverryan committed:
Fixing typo

dunglas commented:
Great! Would allow us to remove some magic regarding the router in API platform as well.

lsmith77 commented:
interesting. would also allow for simple integration of dynamic resources. 

/cc @dbu

sstok commented:
:+1: for this feature!

dupuchba commented:
:+1: really great

weaverryan commented:
The test failure is because FrameworkBundle now depends on a new class inside `symfony/router` that's also included in this PR (but the tests only pull committed code to `symfony/router`). Does anyone know if that test is safe to ignore - or if there's a way to fix it?

Apart from fixing or finding out about the test, this is ready for feedback or voting.

dbu commented on src/Symfony/Component/Routing/Loader/ServiceRouterLoader.php:
```diff
@@ -0,0 +1,78 @@
+/**
+ * A route loader that executes a service to load the routes.
+ *
+ * The depends on the DependencyInjection component
```
typo: "The depends on"

dbu commented on src/Symfony/Component/Routing/Loader/ServiceRouterLoader.php:
```diff
@@ -0,0 +1,78 @@
+        $routeCollection = $service->getRouteCollection($this);
+
+        // make the service file tracked so that if it changes, the cache rebuilds
+        $this->addClassResource(new \ReflectionClass($service), $routeCollection);
```
is it the job of the loader to add possible configuration files to the resources for the cache to invalidate? and how would that work e.g. with routes loaded from a database or some other resource? is there a way to actively invalidate the route cache? probably more of a request for the documentation you write than to go into the phpdoc.

stof replied:
it is indeed the job of the loader to register resources, as the loader knows how it loaded routes.

For routes loaded from the database, you would have to implement a custom resource for that. and it would be a pain until #7230 is implemented.
However, a database loader is probably a very bad idea: the resource-based invalidation of the cache is done only in debug-mode, and you should never put the production website in debug mode (unless you don't care about performance at all of course, but you could as well disable the routing cache to simplify your life if you don't care). So you would have to clear your prod cache explicitly each time the DB changes. This is exactly why the CMF project does not implement a database loader but implements a separate dynamic router instead. This way, the cached router contains only static routes, not dynamic ones, and there is no need to invalidate the routing cache on each DB change.

dbu replied:
thanks. the documentation should explain how to add the resources to help people implementing it correctly. and to avoid people making mistakes or having misconceptions, the discussion of database resources should go into the doc as well. possibly with an example how to actively invalidate the routing cache on relevant changes for people with a specific use case where the CMF dynamic router would be overkill.

mpdude replied:
First: Service-based validation of resources is readily available in #15738, only awaiting review and +1s.

Confession time: We're using router and translator tweaks here that perform cache validation also in production. Yes, you need to consider performance, but if you have only few resources to check (in prod mode), can efficiently query the database for changes, possibly postpone dumping the new router/translator to the termination phase of a request and apply locking patterns (so not every request tries to dump the new routes/translations when the DB changes) it's working fine.

dbu commented:
indeed, good idea!

stof commented on src/Symfony/Component/Routing/Loader/ServiceRouterLoader.php:
```diff
@@ -0,0 +1,78 @@
+    private function addClassResource(\ReflectionClass $class, RouteCollection $collection)
+    {
+        do {
+            $collection->addResource(new FileResource($class->getFileName()));
```
you should check for ``$class->getFileName()`` not being ``false`` (or ``null`` for HHVM) to handle the case of classes extending a class defined in a PHP extension IMO.

weaverryan replied:
This part was stolen from `ContainerBuilder`: https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L275. Is your comment still true? If so - that would be a problem to fix in both places.

weaverryan committed:
Fixing comment typo

xabbuh commented:
@weaverryan The failure for the low tests looks strange. Does it make a difference when you mock the `ContainerInterface` instead?

xabbuh commented:
What I do not understand is the reason to have a container aware routing loader. Wouldn't it be easier if this loader were more like a registry where you could register `RouteLoaderInterface` instances with arbitrary keys (the FrameworkBundle could ship with a compiler pass so that you can just tag your services and they would be registered automatically)?

weaverryan commented:
@xabbuh I understand your thought, but the purpose of this is different. I'm not aiming to make it possible to have RouteLoaderInterface services automatically called because they're registered+tagged. I want to keep loading routes explicit, just add one more way to load routes - via PHP in a real method+service (currently, you can only add routes via PHP in a flat-file).

weaverryan committed:
Fixing test - bad mock

weaverryan commented:
@xabbuh yes - you were right about one of the tests - I've fixed that now.

I believe the last failing test should be ignored: the functional tests in FrameworkBundle are not able to fetch the `router` service because it depends on the `ServiceRouteLoader`, which isn't included in the Symfony's core (since it's part of this PR). I don't think there's way to fix that (someone let me know if I'm wrong), but the tests should pass fine once merged.

Thanks!

xabbuh commented:
@weaverryan I just mean that it would only be possible to use these services in your routing configuration if they have been registered that way before. The workflow would still be the same then with your approach. The only difference is that you would not try to locate the service in the container (the drawback might be that you then won't be able to lazy load the routing loader services though).

weaverryan commented:
@xabbuh Ah, I understand better now. I'd still prefer to keep it this way - it feels logical to me that I should only need to register a class as a service and point to it in the config - the tagging seems unnecessary.

weaverryan committed:
Fixing bad mock of "abstract" class

xabbuh commented:
:+1: (being able to load services lazily and not having to tag your services are good reasons for keeping the container awareness)

weaverryan commented:
Status: Reviewed 

... with some up-votes + "I like this" from core members :)

fabpot commented on src/Symfony/Component/Routing/Loader/ServiceRouterLoader.php:
```diff
@@ -0,0 +1,78 @@
+
+use Symfony\Component\Config\Loader\Loader;
+use Symfony\Component\Config\Resource\FileResource;
+use Symfony\Component\DependencyInjection\ContainerInterface;
```
As it depends on the container, I would have put it in FrameworkBundle instead. At least, that's what we did in the past in such a situation. Not sure what others think about this.

fabpot replied:
Another possibility would be to have an abstract class here that would support any container by adding an abstract protected method like `abstract protected function getService($id)` that then could be specialized for the Symfony SC or Pimple for Silex.

fabpot commented on src/Symfony/Component/Routing/Loader/ServiceRouterLoader.php:
```diff
@@ -0,0 +1,78 @@
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class ServiceRouterLoader extends Loader
```
It we go down the abstract class I've described in my previous comment, this is now more like an `ObjectRouterLoader` as the class would be able to load routes from any object implementing the interface.

fabpot commented:
I've just read the usage in #15948, so this is actually not needed at all. See my comment there (for what I think would a better solution).

weaverryan committed:
Changing this to be an abstract ObjectRouteLoader

weaverryan committed:
Adding a ServiceRouterLoader that works with the DependencyInjection component

weaverryan committed:
Renaming abstract method and going back to the key "service" as the type

weaverryan commented on src/Symfony/Component/Routing/Loader/ObjectRouteLoader.php:
```diff
@@ -0,0 +1,76 @@
+     */
+    public function supports($resource, $type = null)
+    {
+        return 'service' === $type;
```
I was caught between calling this "object" (which would look so generic in a `routing.yml` file) and `service`.

weaverryan commented on src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoader.php:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class ServiceRouterLoader extends ObjectRouteLoader
```
I didn't put this inside `FrameworkBundle` because it's perfectly usable by someone like Drupal. It's really a "Bridge" class, but there's no proper bridge to put it into.

weaverryan commented on src/Symfony/Component/Routing/Loader/ObjectRouteLoader.php:
```diff
@@ -0,0 +1,76 @@
+     * @param string $id
+     * @return RouteLoaderInterface
+     */
+    abstract protected function getRouteLoaderService($id);
```
We could remove the RouteLoaderInterface and instead have this method return a callable.

The difference would then be that we'd probably need to include the method name in the "resource" when using this:

```yml
admin_routes:
    resource: 'admin_route_loader::loadRoutes'
    type: service
```

That would solve the need for the RouteLoaderInterface on the "Micro" kernel.

weaverryan committed:
fabbot

weaverryan commented:
Adding a comment that got covered up:

We could remove the `RouteLoaderInterface` and instead have this method return a callable.

The difference would then be that we'd probably need to include the method name in the "resource" when using this:

```yml
admin_routes:
    resource: 'admin_route_loader::loadRoutes'
    type: service
```

That would solve the need for the RouteLoaderInterface on the "Micro" kernel. It's also consistent with how event listeners, for example, work.

fabpot commented:
@weaverryan Indeed, that would be much better. IIRC, when using a service name, the convention is to use only one colon: `admin_router_loader:loadRoutes`

fabpot commented:
#15990 has been updated to work with the new `service` loader.

weaverryan commented:
@fabpot I've already done work locally on this PR for the `kernel:methodName` syntax - I'll be able to push it within the next 2 hours.

weaverryan committed:
Updating ObjectRouteLoader to use a service_id:methodName format

weaverryan commented:
@fabpot a little later, but I just updated the PR:

This uses the `service_name:method` notation and is enforced in the `ObjectRouteLoader`, instead of allowing for more generic usage, where sub-classes could resolve strings of any format. I was trying to limit it to work just one way. For example, if you're not worried about the cached container and want to load routes via some "Callable", you cannot do it via the `ObjectLoader`: use the `ClosureLoader`.

ping also @symfony/deciders

aitboudad commented:
:+1: , can you update the [CHANGELOG](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Routing/CHANGELOG.md).

weaverryan committed:
updating CHANGELOG

weaverryan commented:
done!

fabpot commented on src/Symfony/Component/Routing/Loader/ObjectRouteLoader.php:
```diff
@@ -0,0 +1,95 @@
+        $loaderObject = $this->getServiceObject($serviceString);
+
+        if (!is_object($loaderObject)) {
+            throw new \LogicException(sprintf('%s:getServiceObject() must return an object: %s returned', get_class($this), $loaderObject));
```
Here, `$loaderObject` can be anything; I would return the type instead.

weaverryan replied:
very good catch - that's what I had meant to do. Fixed now!

fabpot commented on src/Symfony/Component/Routing/Loader/ObjectRouteLoader.php:
```diff
@@ -0,0 +1,95 @@
+            throw new \LogicException(sprintf('The %s::%s method must return a RouteCollection: %s returned', get_class($loaderObject), $method, $type));
+        }
+
+        // make the service file tracked so that if it changes, the cache rebuilds
```
change**d**

fabpot commented:
I made a small comment; :+1:

weaverryan committed:
Fixing missing gettype

weaverryan commented:
comment addressed!

fabpot commented:
I think this one is ready now, ping @symfony/deciders

fabpot commented:
Thank you @weaverryan.
